### PR TITLE
Legacy address

### LINKED ISSUE #44: Update crypro library with Bitcoin operations
For Bitcoin integration, we need to add all crypto-related functions, that are needed for Bitcoin but are not used in Skycoin. We need to use the functions from [trezor](https://github.com/trezor/trezor-mcu) implementation to be sure, that they work well

- [x] Investigate which functions should be added to our crypto library
- [ ] Add them to our library
- [ ] Write tests to be sure, that they are working well.

### PR DESCRIPTION
**Pull request by Macrosrider**
Fixes #44, #45 	

 Changes:	
- Added BTC legacy address generation	

 Does this change need to mentioned in CHANGELOG.md?	
yes

 Requires testing	
yes

 Comments about testing , should you have some

### CODE REVIEW

File: skycoin-api/skycoin_crypto.c [hunk 1]
```diff
@@ -345,6 +345,34 @@ int skycoin_address_from_pubkey(const uint8_t* pubkey, char* b58address, size_t*
     return 0;
 }
 
+int bitcoin_address_from_pubkey(const uint8_t* pubkey, char* b58address, size_t* size_b58address){
```
Comment (Midren) [inline skycoin-api/skycoin_crypto.c]: This function should not be in the skycoin_crypto.c file. Skycoin and Bitcoin code should be divided in different files

File: skycoin-api/test_skycoin_crypto.c [hunk 2]
```diff
@@ -668,6 +668,27 @@ START_TEST(test_skycoin_address_from_pubkey)
 }
 END_TEST
 
+START_TEST(test_bitcoin_address_from_pubkey){
```
Comment (Midren) [inline skycoin-api/test_skycoin_crypto.c:673]: New file for bitcoin tests also should be done.

File: tiny-firmware/Makefile [hunk 3]
```diff
@@ -36,6 +36,8 @@ OBJS += $(FIRMWARE_SRCS:.c=.o)
 
 OBJS += protob/c/messages.pb.o
 OBJS += protob/c/types.pb.o
+OBJS += protob/c/bitcoin_messages.pb.o
+OBJS += protob/c/bitcoin_types.pb.o
```
Comment (Midren) [inline tiny-firmware/Makefile]: Rewrite here makefile, so it will include all files from `protob/c`, so nextly we don't need to change Makefile, if add new files in protobuf.

File: tiny-firmware/firmware/fsm.c [hunk 4]
```diff
@@ -633,4 +633,34 @@ void fsm_msgTxAck(TxAck *msg) {
     }
     layoutHome();
     return;
-}
\ No newline at end of file
+}
+
+void fsm_msgBitcoinTxAck(BitcoinTxAck *msg) {
```
Comment (Midren) [inline tiny-firmware/firmware/fsm.c:638]: This PR adds an address functionality, this should be removed and added, as separate PR

Comment (Midren) [reply tiny-firmware/firmware/fsm.c:638]: Also, don't forget to put this function to in a new file bitcoin_fsm.c

File: tiny-firmware/tests/test_fsm_bitcoin.c [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+void setBitcoinPassphrase(char *passphrase) {
+    ApplySettings msg = ApplySettings_init_zero;
+    msg.has_use_passphrase = true;
+    msg.use_passphrase = true;
+    ck_assert_int_eq(msgApplySettingsImpl(&msg), ErrOk);
+    session_cachePassphrase(passphrase);
+}
+
+START_TEST(test_msgBitcoinAddressesAllEmptyPassphrase)
+        {
```
Comment (Midren) [inline tiny-firmware/tests/test_fsm_bitcoin.c:136]: Problems with a linter. Fix ths indentations

File: tiny-firmware/firmware/messages.c [hunk 6]
```diff
@@ -268,6 +270,8 @@ void msg_read_tiny(const uint8_t* buf, int len)
     case MessageType_MessageType_Initialize:
         fields = Initialize_fields;
         break;
+    case MessageType_MessageType_BitcoinTxAck:
```
Comment (Midren) [inline tiny-firmware/firmware/messages.c]: Transactions are not part of this PR.

File: tiny-firmware/firmware/fsm_impl.c [hunk 7]
```diff
@@ -250,7 +254,7 @@ fsm_getKeyPairAtIndex(uint32_t nbAddress, uint8_t *pubkey, uint8_t *seckey, Resp
         return ErrFailed;
     }
     if (respSkycoinAddress != NULL && start_index == 0) {
-        if (!skycoin_address_from_pubkey(pubkey, respSkycoinAddress->addresses[0], &size_address)) {
+        if (!(address_from_pubkey)(pubkey, respSkycoinAddress->addresses[0], &size_address)) {
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_impl.c]: Parentheses are redundant here.

File: tiny-firmware/firmware/fsm_impl.c [hunk 8]
```diff
@@ -270,7 +274,7 @@ fsm_getKeyPairAtIndex(uint32_t nbAddress, uint8_t *pubkey, uint8_t *seckey, Resp
         seed[32] = 0;
         if (respSkycoinAddress != NULL && ((i + 1) >= start_index)) {
             size_address = 36;
-            if (!skycoin_address_from_pubkey(pubkey, respSkycoinAddress->addresses[respSkycoinAddress->addresses_count],
+            if (!(address_from_pubkey)(pubkey, respSkycoinAddress->addresses[respSkycoinAddress->addresses_count],
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_impl.c]: The same as above.

File: tiny-firmware/firmware/fsm_impl.c [hunk 9]
```diff
@@ -213,7 +215,9 @@ ErrCode_t msgSignTransactionMessageImpl(uint8_t *message_digest, uint32_t index,
     uint8_t pubkey[SKYCOIN_PUBKEY_LEN] = {0};
     uint8_t seckey[SKYCOIN_SECKEY_LEN] = {0};
     uint8_t signature[SKYCOIN_SIG_LEN];
-    ErrCode_t res = fsm_getKeyPairAtIndex(1, pubkey, seckey, NULL, index);
+    int (*address_from_pubkey)(const uint8_t*, char*, size_t*);
+    address_from_pubkey = &skycoin_address_from_pubkey;
+    ErrCode_t res = fsm_getKeyPairAtIndex(1, pubkey, seckey, NULL, index, address_from_pubkey);
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_impl.c]: You can omit creating `address_from_pubkey` variable and directly path `&skycoin_address_from_pubkey` to fsm_getKeyPairAtIndex

File: tiny-firmware/firmware/fsm_impl.c [hunk 10]
```diff
@@ -738,4 +742,10 @@ ErrCode_t msgTxAckImpl(TxAck *msg, TxRequest *resp) {
     if (resp->request_type == TxRequest_RequestType_TXFINISHED)
         TxSignCtx_Destroy(ctx);
     return ErrOk;
-}
\ No newline at end of file
+}
+
+ErrCode_t msgBitcoinTxAckImpl(BitcoinTxAck *msg, TxRequest *resp){
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_impl.c:768]: Bitcoin Transactions are not part of these PR.

File: tiny-firmware/firmware/fsm_skycoin.c [hunk 11]
```diff
@@ -86,7 +87,9 @@ void fsm_msgSkycoinSignMessage(SkycoinSignMessage *msg) {
     ResponseSkycoinAddress respAddr;
     uint8_t seckey[32] = {0};
     uint8_t pubkey[33] = {0};
-    ErrCode_t err = fsm_getKeyPairAtIndex(1, pubkey, seckey, &respAddr, msg->address_n);
+    int (*address_from_pubkey)(const uint8_t*, char*, size_t*);
+    address_from_pubkey = &skycoin_address_from_pubkey;
+    ErrCode_t err = fsm_getKeyPairAtIndex(1, pubkey, seckey, &respAddr, msg->address_n, address_from_pubkey);
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_skycoin.c]: You can omit creating `address_from_pubkey` variable and directly path `&skycoin_address_from_pubkey` to fsm_getKeyPairAtIndex.

File: tiny-firmware/firmware/fsm_skycoin_impl.c [hunk 12]
```diff
@@ -107,7 +107,9 @@ ErrCode_t msgSkycoinSignMessageImpl(SkycoinSignMessage *msg, ResponseSkycoinSign
     uint8_t seckey[SKYCOIN_SECKEY_LEN] = {0};
     uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
     uint8_t signature[SKYCOIN_SIG_LEN];
-    if (fsm_getKeyPairAtIndex(1, pubkey, seckey, NULL, msg->address_n) != ErrOk) {
+    int (*address_from_pubkey)(const uint8_t*, char*, size_t*);
+    address_from_pubkey = &skycoin_address_from_pubkey;
+    if (fsm_getKeyPairAtIndex(1, pubkey, seckey, NULL, msg->address_n, address_from_pubkey) != ErrOk) {
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_skycoin_impl.c]: You can omit creating `address_from_pubkey` variable and directly path `&skycoin_address_from_pubkey` to fsm_getKeyPairAtIndex.

File: tiny-firmware/firmware/fsm_skycoin_impl.c [hunk 13]
```diff
@@ -146,7 +150,7 @@ ErrCode_t msgSkycoinAddressImpl(SkycoinAddress *msg, ResponseSkycoinAddress *res
         return ErrMnemonicRequired;
     }
 
-    if (fsm_getKeyPairAtIndex(msg->address_n, pubkey, seckey, resp, start_index) != ErrOk) {
+    if (fsm_getKeyPairAtIndex(msg->address_n, pubkey, seckey, resp, start_index, address_from_pubkey) != ErrOk) {
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_skycoin_impl.c]: You can omit creating `address_from_pubkey` variable and directly path `&skycoin_address_from_pubkey` to fsm_getKeyPairAtIndex.

File: tiny-firmware/firmware/fsm_skycoin_impl.c [hunk 14]
```diff
@@ -209,7 +213,9 @@ msgTransactionSignImpl(TransactionSign *msg, ErrCode_t (*funcConfirmTxn)(char *,
             uint8_t seckey[32] = {0};
             size_t size_address = 36;
             char address[36] = {0};
-            ErrCode_t ret = fsm_getKeyPairAtIndex(1, pubkey, seckey, NULL, msg->transactionOut[i].address_index);
+            int (*address_from_pubkey)(const uint8_t*, char*, size_t*);
+            address_from_pubkey = &skycoin_address_from_pubkey;
+            ErrCode_t ret = fsm_getKeyPairAtIndex(1, pubkey, seckey, NULL, msg->transactionOut[i].address_index, address_from_pubkey);
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_skycoin_impl.c]: You can omit creating `address_from_pubkey` variable and directly path `&skycoin_address_from_pubkey` to fsm_getKeyPairAtIndex.

File: tiny-firmware/firmware/fsm_bitcoin_impl.c [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    if (msg->address_n > 99) {
+        return ErrTooManyAddresses;
+    }
+
+    if (storage_hasMnemonic() == false) {
+        return ErrMnemonicRequired;
+    }
+
+    if (fsm_getKeyPairAtIndex(msg->address_n, pubkey, seckey, resp, start_index, address_from_pubkey) != ErrOk) {
```
Comment (Midren) [inline tiny-firmware/firmware/fsm_bitcoin_impl.c]: You can omit creating `address_from_pubkey` variable and directly path `&bitcoin_address_from_pubkey` to fsm_getKeyPairAtIndex.

File: skycoin-api/skycoin_constants.h [hunk 16]
```diff
@@ -5,5 +5,7 @@
 #define SKYCOIN_SECKEY_LEN 32
 #define SKYCOIN_SIG_LEN 65
 #define SKYCOIN_ADDRESS_CHECKSUM_LENGTH 4
+#define SKY_COIN 0
+#define BIT_COIN 1
```
Comment (Midren) [inline skycoin-api/skycoin_constants.h]: These defines are not used, why do we need them?

File: skycoin-api/skycoin_crypto.c [hunk 17]
```diff
@@ -345,6 +345,34 @@ int skycoin_address_from_pubkey(const uint8_t* pubkey, char* b58address, size_t*
...
+    const curve_info* curve = get_curve_by_name(SECP256K1_NAME);
+
+    if (!pubkey_is_valid(curve->params, pubkey)) {
+        return 0;
+    }
+
+    uint8_t address[RIPEMD160_DIGEST_LENGTH + 1 + 4] = {0};
```
Comment (Midren) [inline skycoin-api/skycoin_crypto.c]: Add similar as SKYCOIN_ADDRESS_CHECKSUM_LENGTH constant for Bitcoin checksum and use it instead of 4.

File: skycoin-api/skycoin_crypto.c [hunk 18]
```diff
@@ -345,6 +345,34 @@ int skycoin_address_from_pubkey(const uint8_t* pubkey, char* b58address, size_t*
...
+    uint8_t r1[SHA256_DIGEST_LENGTH] = {0};
+    uint8_t r2[SHA256_DIGEST_LENGTH] = {0};
+
+    sha256sum(pubkey, r1, SKYCOIN_PUBKEY_LEN);
+    ripemd160(r1, SHA256_DIGEST_LENGTH, address);
+
+    uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
+    memmove(address+1, address, RIPEMD160_DIGEST_LENGTH + 4);
```
Comment (Midren) [inline skycoin-api/skycoin_crypto.c]: The same as above.

File: skycoin-api/skycoin_crypto.c [hunk 19]
```diff
@@ -345,6 +345,34 @@ int skycoin_address_from_pubkey(const uint8_t* pubkey, char* b58address, size_t*
...
+    address[0] = 0; // version byte
+
+    sha256sum(address, digest, RIPEMD160_DIGEST_LENGTH + 1);
+    sha256sum(digest, r2, sizeof(digest));
+    memcpy(&address[RIPEMD160_DIGEST_LENGTH + 1], r2, SKYCOIN_ADDRESS_CHECKSUM_LENGTH);
```
Comment (Midren) [inline skycoin-api/skycoin_crypto.c]: Replace `SKYCOIN_ADDRESS_CHECKSUM_LENGTH` with bitcoin analog.

File: skycoin-api/skycoin_crypto.c [hunk 20]
```diff
@@ -345,6 +345,34 @@ int skycoin_address_from_pubkey(const uint8_t* pubkey, char* b58address, size_t*
...
+
+    if (!pubkey_is_valid(curve->params, pubkey)) {
+        return 0;
+    }
+
+    uint8_t address[RIPEMD160_DIGEST_LENGTH + 1 + 4] = {0};
+    uint8_t r1[SHA256_DIGEST_LENGTH] = {0};
+    uint8_t r2[SHA256_DIGEST_LENGTH] = {0};
+
+    sha256sum(pubkey, r1, SKYCOIN_PUBKEY_LEN);
```
Comment (Midren) [inline skycoin-api/skycoin_crypto.c]: Add `BITCOIN_PUBKEY_LEN` constant analog and replace `SKYCOIN_PUBKEY_LEN`

**Review by Midren — CHANGES_REQUESTED** [review]

File: tiny-firmware/firmware/fsm.c (hunk 4 above)
Comment (Macrosrider) [reply tiny-firmware/firmware/fsm.c:638]: It should be declared, because without it I got the following error:
```skycoinProject/hardware-wallet/tiny-firmware//protob/c/messages_map.h:34:110: error: 'fsm_msgBitcoinTxAck' undeclared here (not in a function); did you mean '_BitcoinTxAck'?```

**Comment by Midren** [comment]
As for me, we need to rename skycoin-api to skywallet-crypto, as from now it will have functionality not only for Skycoin, and then we will import only one library for the firmware. What do you think @jdknives ?

**Comment by Midren** [comment]
@Macrosrider Rename skycoin-api folder to skywallet-crypto
